build/libribasim/src/libribasim.jl

```diff
@@ -71,10 +91,61 @@ Base.@ccallable function update_until(time::Cdouble)::Cint
     end
 end
 
-# TODO should receive a double that needs to get updated
-Base.@ccallable function get_current_time()::Cint
+Base.@ccallable function get_current_time(time::Ptr{Cdouble})::Cint
+    @try_c begin
+        t = BMI.get_current_time(model)
+        unsafe_store!(time, t)
+    end
+end
+
+Base.@ccallable function get_var_type(name::Cstring, var_type::Cstring)::Cint
+    @try_c begin
+        value = BMI.get_value_ptr(model, unsafe_string(name))
+        dtype = if value isa Vector
+            julia_type_to_numpy(eltype(value))
+        elseif value isa Number
+            julia_type_to_numpy(typeof(value))
+        else
+            error("Unsupported value type $(typeof(value))")
+        end
+
+        var_type_ptr = pointer(var_type)
+        for (i, char) in enumerate(ascii(dtype))
+            unsafe_store!(var_type_ptr, char, i)
+        end
+        unsafe_store!(var_type_ptr, '\0', length(dtype) + 1)
+    end
+end
+
+Base.@ccallable function get_var_rank(name::Cstring, rank::Ptr{Cdouble})::Cint
     @try_c begin
-        BMI.get_current_time(model)
+        value = BMI.get_value_ptr(model, unsafe_string(name))
+        n = ndims(value)
+        unsafe_store!(rank, n)
+    end
+end
+
+Base.@ccallable function get_value_ptr(name::Cstring, value_ptr::Ptr{Ptr{Cvoid}})::Cint
+    @try_c begin
+        value = BMI.get_value_ptr(model, unsafe_string(name))
+        n = length(value)
+        core_ptr = Base.unsafe_convert(Ptr{Ptr{Cvoid}}, value)
+        unsafe_copyto!(value_ptr, core_ptr, length(value))
+    end
+end
+
+Base.@ccallable function get_value_ptr_double(
+    name::Cstring,
+    value_ptr::Ptr{Ptr{Cvoid}},
+)::Cint
+    get_value_ptr(name, value_ptr)
+end
+
+function julia_type_to_numpy(type)::String
```

Comment:
Our C API is numpy specific? I'm not sure how BMI would work with other languages when one can retrieve the dtype.